fedimint-server/src/config.rs

```diff
@@ -166,17 +181,30 @@ pub type DynModuleInit = Arc<dyn ModuleInit + Send + Sync>;
 // TODO: turn all registeries into `Registry<T>(BTreeMap<&'static str, T))` + type alias
 #[autoimpl(Deref using self.0)]
 #[derive(Clone)]
-pub struct ModuleInitRegistry(BTreeMap<&'static str, DynModuleInit>);
+pub struct ModuleInitRegistry(BTreeMap<ModuleKind, DynModuleInit>);
 
-impl<const N: usize> From<[(&'static str, DynModuleInit); N]> for ModuleInitRegistry {
-    fn from(value: [(&'static str, DynModuleInit); N]) -> Self {
-        Self(BTreeMap::from(value))
+impl From<Vec<DynModuleInit>> for ModuleInitRegistry {
+    fn from(value: Vec<DynModuleInit>) -> Self {
+        Self(BTreeMap::from_iter(
+            value.into_iter().map(|i| (i.module_kind(), i)),
+        ))
     }
 }
 
 impl ModuleInitRegistry {
-    pub fn decoders(&self) -> ModuleDecoderRegistry {
-        ModuleDecoderRegistry::from_iter(self.0.values().map(|v| v.decoder()))
+    pub fn decoders<'a>(
+        &self,
+        module_kinds: impl Iterator<Item = (ModuleInstanceId, &'a ModuleKind)>,
+    ) -> anyhow::Result<ModuleDecoderRegistry> {
+        let mut modules = BTreeMap::new();
+        for (id, kind) in module_kinds {
+            let Some(init) = self.0.get(kind) else {
+                anyhow::bail!("Detected configuration for unsupported module kind: {kind:?}")
+            };
```

Comment:
First time I see this syntax used after it became available :heart_eyes:

Reply:
It's nice, but great you reminded me. I actually have a Display on it already.